cpp/src/copying/contiguous_split.cu

```diff
@@ -14,465 +14,923 @@
+  CUDA_TRY(cudaStreamSynchronize(stream));
+
+  // allocate output partition buffers
+  std::vector<rmm::device_buffer> out_buffers;
+  out_buffers.reserve(num_partitions);
+  std::transform(h_buf_sizes,
+                 h_buf_sizes + num_partitions,
+                 std::back_inserter(out_buffers),
+                 [stream, mr](size_t bytes) {
+                   return rmm::device_buffer{bytes, stream, mr};
                  });
 
+  // clang-format off
+  // packed block of memory 3. pointers to source and destination buffers (and stack space on the
+  // gpu for offset computation)
+  size_t const src_bufs_size = cudf::util::round_up_safe(num_src_bufs * sizeof(uint8_t*), split_align);
+  size_t const dst_bufs_size = cudf::util::round_up_safe(num_partitions * sizeof(uint8_t*), split_align);
+  // host-side
+  std::vector<uint8_t> h_src_and_dst_buffers(src_bufs_size + dst_bufs_size);
+  uint8_t** h_src_bufs = reinterpret_cast<uint8_t**>(h_src_and_dst_buffers.data());
+  uint8_t** h_dst_bufs = reinterpret_cast<uint8_t**>(h_src_and_dst_buffers.data() + src_bufs_size);
+  // device-side
+  rmm::device_buffer d_src_and_dst_buffers(src_bufs_size + dst_bufs_size + offset_stack_size, stream, rmm::mr::get_current_device_resource());
+  uint8_t** d_src_bufs = reinterpret_cast<uint8_t**>(d_src_and_dst_buffers.data());
+  uint8_t** d_dst_bufs = reinterpret_cast<uint8_t**>(reinterpret_cast<uint8_t*>(d_src_and_dst_buffers.data()) + src_bufs_size);
+  // clang-format on
+
+  // setup src buffers
+  setup_src_buf_data(input.begin(), input.end(), h_src_bufs);
+
+  // setup dst buffers
+  std::transform(out_buffers.begin(), out_buffers.end(), h_dst_bufs, [](auto& buf) {
+    return static_cast<uint8_t*>(buf.data());
+  });
+
+  // HtoD src and dest buffers
+  CUDA_TRY(cudaMemcpyAsync(
+    d_src_bufs, h_src_bufs, src_bufs_size + dst_bufs_size, cudaMemcpyHostToDevice, stream));
+
+  // copy.  1 block per buffer
+  {
+    constexpr int block_size = 512;
+    copy_partition<<<num_bufs, block_size, 0, stream>>>(
+      num_src_bufs, num_partitions, d_src_bufs, d_dst_bufs, d_dst_buf_info);
+  }
+
+  // build the output.
+  std::vector<contiguous_split_result> result;
+  result.reserve(num_partitions);
+  std::vector<column_view> cols;
+  cols.reserve(num_root_columns);
+  auto cur_dst_buf_info = h_dst_buf_info;
+  for (size_t idx = 0; idx < num_partitions; idx++) {
+    cur_dst_buf_info = build_output_columns(
+      input.begin(), input.end(), cur_dst_buf_info, std::back_inserter(cols), h_dst_bufs[idx]);
+    result.push_back(contiguous_split_result{
+      cudf::table_view{cols}, std::make_unique<rmm::device_buffer>(std::move(out_buffers[idx]))});
+    cols.clear();
+  }
+
+  // overlap the actual copy with the work of constructing the output columns. this can be a
+  // non-trivial savings because of the sheer number of output views.
+  CUDA_TRY(cudaStreamSynchronize(stream));
```

Comment:
Wait, why do you need to synchronize here? If the user passes a non-default stream, they should know that any device-side result must be accessed using stream ordering semantics.

Reply:
Because there were async copies from host stack data. The copies may not have completed by the time the function returns at which point the host data is no longer valid.

Reply:
Maybe a comment to that affect. Current comment talks about overlap, but a sync doesn't cause overlap, it ends it.

Reply:
The existing comment was just to note that the work of doing the `build_output_columns` should be overlapped with the execution of the kernel (constructing those result columns can be pretty expensive on the CPU).  But Jake's comment also applies :  the HtoD memcpy immediately before the copy kernel could still be in-flight. I'll update the comment.